### ISSUE TITLE
Cytofkit does not include all FCS files

### ISSUE THREAD

**Issue by taxkourel** [JinmiaoChenLab/cytofkit#47]
Hello,
Trying to run an analysis and although I select all my FCS files in the "FCS files" folder only some of them are included. I have R3.5.1. I tried to shorten the file name and then all get included but including the entire filename would make things much easier.
Any idea how to bypass this?
Thank you

**Comment by SamGG** [comment]
Hi,
Could you specify your OS (Linux, Win, Mac)? Under Windows, I would recommand to move the working  folder to the root of your disk (ie C:\data\my_working_folder). Tell us if the folder is located on your disk or a network share.
Best.

**Comment by taxkourel** [comment]
I have windows 7 and files are on my desktop’s disk.

It is the weirdest. It think something in the existing file name is confusing it.
I tried with “artificially long” file names (see below) and it could upload them irrespective of folder location. When I try my existing file names (equal length overall) it gets confused.
For e.g.:

Current file name format is

EXP10_Sample01_Diagnosis_Name_SampleID_date.fcs

Diagnosis is anywhere from 2-6 characters
Name 2-7 characters
Sample ID 10 alphanumeric
Date 8 numeric

When I shorten to “EXP10_Sample01_.fcs” format they go through
When I lengthen to “EXP10_this is a very very very very very very very very very long file name.fcs” it also goes through!
Maybe it is the “_”?
If you have any ideas please let me know since filenames are quite informative an d help during downstream analyses.
Thank you for your quick reply!
Tax

**Comment by SamGG** [comment]
I don't think the '_' may be an issue. Looking at the code, the characters '{' or '}' could be an issue. In the following lines of R, I tried to mimic cytofkit_GUI function in selecting FCS files.
Alternatively, it seems that you can select the directory of the FCS files, and I guess all FCS will be included in the analysis without selecting them one by one.
Last option is to setup the proper call to the cytofkit() function.
Copy and run the following lines, it should help you understand where things go wrong.
```
# select the file names using the TCLTK package in the same way as cytofkit
# you have to go through the disk tree up to the directory where the FCS files are
fnames <- ""
fnames <- tcltk::tk_choose.files(default = ".", caption = "Select FCS files", multi = TRUE, filters = matrix(c("{fcs files}", "{.fcs}"), 1, 2))
fnames # selected files

# remove empty .fcs files (?)
grepl(paste0(.Platform$file.sep, "fcs$"), fnames)
fnames <- fnames[!(grepl(paste0(.Platform$file.sep, "fcs$"), fnames))]
fnames # filtered files
# and store them in a value of the interface
tclvalue.fcsFile <- paste(fnames, collapse = "}{")

# extract file names later on
strsplit(tclvalue.fcsFile, "}{", fixed = TRUE)[[1]]
# do you get the files you selected originally?
```

**Comment by SamGG** [comment]
In case of it could help this problem
Install latest version of flowCore (1.47.5) from github and report if the problem is still there.
```
install.packages("devtools")
devtools::install_github("RGLab/flowCore")
```

**Comment by CNicholasMDA** [comment]
Hi Sam and Tax!
Have you had any success combating this problem?  If so, what was the best solution?
Thank you!
-Courtney

**Comment by taxkourel** [comment]
I did not install the most uptodate version of flowcore which I think has solved it for most people. I had very long file names and when I simplified the name it worked (no idea why). I have not used cytofkit recently though ☹

**Comment by CNicholasMDA** [comment]
Thanks, Tax.  That helps.  Those are two solutions that I can consider. 

Best wishes with your research.  Please let me know if I can ever help in the future.

-Courtney

**Comment by CNicholasMDA** [comment]
Hi again Tax and Sam,
I just learned something new and disturbing about our fcs problem.  I analyzed 118 fcs files in cyTOFKIT, and the resulting analysis had 118 files.  Lookin' good so far.  *Then*, I decide to rename (called "regroup") my samples to make it easier to interpret the patient ID #s and experimental timepoints.  After I do this task, I notice that I have lost ~18 fcs files (or, ~3 patients with 5 timepoints each) from my dataset.

So, there might be multiple issues overlying each other, not the least of which is ... it's advised that you don't change or re-group your sample names until this bug is fixed.

Do not hesitate if you have a question or comment about what i've proposed above.
happy computing,
Courtney

**Comment by SamGG** [comment]
Hi Courtney,
What is the version of flowCore?

**Comment by CNicholasMDA** [comment]
Hi Sam,
Currently, I am running  flowCore 1.44.2. However, I had the same issue when I installed 1.47.5.  

This issue is, I think, related to another issue I opened (#49).  You are welcome to take a look at the evidence I posted there.

Any thoughts or suggestions you may have are welcome and appreciated.
-Courtney

**Comment by SamGG** [comment]
Yep, I already followed all your issue submissions. Sorry to insist, but when you installed flowCore 1.47.5, you are absolutely sure you really got flowCore 1.47.5. I mean did you a packageVersion("flowCore") after having the installation to be be certain that this version is correctly installed?

**Comment by CNicholasMDA** [comment]
If I remember correctly, I indeed did just that.  To be clear, I didn't lose .fcs files until *after* I re-grouped (re-named) the files within the cyTOFKIT GUI.

**Comment by SamGG** [comment]
I think the grouping files described here is certainly a distinct issue from the issue #49 and the issue you opened in Premessa. I already worked with Premessa and Cytofcore without any issue. But as Premessa is writing a FCS file at the end of the process, I would run Premessa with flowCore 1.47. Same for issue #49 althougth I don't know if the renamed markers are stored to FCS files before processing. HTH

**Comment by CNicholasMDA** [comment]
Ok!  I'll consider it.  Thank you!
